lib/ecto/query/planner.ex

```diff
@@ -434,60 +434,69 @@ defmodule Ecto.Query.Planner do
 
   defp plan_joins([%JoinExpr{assoc: {ix, assoc}, qual: qual, on: on, prefix: prefix} = join|t],
                      query, joins, sources, tail_sources, counter, offset, adapter) do
-    schema = schema_for_association_join!(query, join, Enum.fetch!(Enum.reverse(sources), ix))
-    refl = schema.__schema__(:association, assoc)
 
-    unless refl do
-      error! query, join, "could not find association `#{assoc}` on schema #{inspect schema}"
-    end
+    with r_sources = Enum.reverse(sources),
+         true <- is_integer(ix),
+         {:ok, source} <- Enum.fetch(r_sources, ix) do
+
```

Comment:
Instead using `with`, you could write a private function that validates & fetches ix or raises otherwise, as you're raising on else anyway

Reply:
Yes, agreed. Instead of nesting all code, you can call `validate_sources!` and keep the logic only in there. In any case, great PR! :+1:

Reply:
What about `fetch_source!(sources, ix)` that actually does this with inside and just returns `source`? It would be used like: `source = fetch_source!(sources, ix)`

Reply:
Sounds good!

Reply:
Refactored and added tests :)